### ISSUE TITLE
Problems compiling  Icinga2 on macOS

### ISSUE THREAD

**Issue by henningkessler** [Icinga/icinga2#8484]
## Describe the bug

compiling icinga2 on macOS (at least on 10.14. and 10.15.) fails with:
```
Scanning dependencies of target base
[  7%] Building CXX object lib/base/CMakeFiles/base.dir/application.cpp.o
[  7%] Building CXX object lib/base/CMakeFiles/base.dir/application-version.cpp.o
[  7%] Building CXX object lib/base/CMakeFiles/base.dir/application-environment.cpp.o
[  8%] Building CXX object lib/base/CMakeFiles/base.dir/array.cpp.o
In file included from /Users/virt/Desktop/icinga2/lib/base/application-version.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/application.hpp:6:
In file included from /Users/virt/Desktop/icinga2/lib/base/i2-base.hpp:38:
In file included from /usr/local/include/boost/config.hpp:44:
/usr/local/include/boost/config/detail/select_stdlib_config.hpp:24:14: warning: non-portable path to file '<VERSION>'; specified path differs in case from file name on disk [-Wnonportable-include-path]
#    include <version>
             ^~~~~~~~~
             <VERSION>
In file included from /Users/virt/Desktop/icinga2/lib/base/application-version.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/application.hpp:6:
In file included from /Users/virt/Desktop/icinga2/lib/base/i2-base.hpp:38:
In file included from /usr/local/include/boost/config.hpp:44:
In file included from /usr/local/include/boost/config/detail/select_stdlib_config.hpp:24:
/Users/virt/Desktop/icinga2/version:1:1: error: unknown type name 'Version'
Version: 2.12.0
^
/Users/virt/Desktop/icinga2/version:1:8: error: expected unqualified-id
Version: 2.12.0
       ^
In file included from /Users/virt/Desktop/icinga2/lib/base/application-version.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/application.hpp:7:
In file included from /Users/virt/Desktop/icinga2/debug/lib/base/application-ti.hpp:9:
In file included from /usr/local/include/boost/signals2.hpp:19:
In file included from /usr/local/include/boost/signals2/signal.hpp:22:
In file included from /usr/local/include/boost/signals2/connection.hpp:22:
In file included from /usr/local/include/boost/signals2/detail/null_output_iterator.hpp:14:
/usr/local/include/boost/function_output_iterator.hpp:14:1: warning: This header is deprecated. Use <boost/iterator/function_output_iterator.hpp> instead. [-W#pragma-messages]
BOOST_HEADER_DEPRECATED("<boost/iterator/function_output_iterator.hpp>")
^
/usr/local/include/boost/config/header_deprecated.hpp:23:37: note: expanded from macro 'BOOST_HEADER_DEPRECATED'
# define BOOST_HEADER_DEPRECATED(a) BOOST_PRAGMA_MESSAGE("This header is deprecated. Use " a " instead.")
                                    ^
/usr/local/include/boost/config/pragma_message.hpp:24:34: note: expanded from macro 'BOOST_PRAGMA_MESSAGE'
# define BOOST_PRAGMA_MESSAGE(x) _Pragma(BOOST_STRINGIZE(message(x)))
                                 ^
<scratch space>:21:2: note: expanded from here
 message("This header is deprecated. Use " "<boost/iterator/function_output_iterator.hpp>" " instead.")
 ^
2 warnings and 2 errors generated.
make[2]: *** [lib/base/CMakeFiles/base.dir/application-version.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /Users/virt/Desktop/icinga2/lib/base/application-environment.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/application.hpp:6:
In file included from /Users/virt/Desktop/icinga2/lib/base/i2-base.hpp:38:
In file included from /usr/local/include/boost/config.hpp:44:
/usr/local/include/boost/config/detail/select_stdlib_config.hpp:24:14: warning: non-portable path to file '<VERSION>'; specified path differs in case from file name on disk [-Wnonportable-include-path]
#    include <version>
             ^~~~~~~~~
             <VERSION>
In file included from /Users/virt/Desktop/icinga2/lib/base/application-environment.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/application.hpp:6:
In file included from /Users/virt/Desktop/icinga2/lib/base/i2-base.hpp:38:
In file included from /usr/local/include/boost/config.hpp:44:
In file included from /usr/local/include/boost/config/detail/select_stdlib_config.hpp:24:
/Users/virt/Desktop/icinga2/version:1:1: error: unknown type name 'Version'
Version: 2.12.0
^
/Users/virt/Desktop/icinga2/version:1:8: error: expected unqualified-id
Version: 2.12.0
       ^
In file included from /Users/virt/Desktop/icinga2/lib/base/application-environment.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/application.hpp:7:
In file included from /Users/virt/Desktop/icinga2/debug/lib/base/application-ti.hpp:9:
In file included from /usr/local/include/boost/signals2.hpp:19:
In file included from /usr/local/include/boost/signals2/signal.hpp:22:
In file included from /usr/local/include/boost/signals2/connection.hpp:22:
In file included from /usr/local/include/boost/signals2/detail/null_output_iterator.hpp:14:
/usr/local/include/boost/function_output_iterator.hpp:14:1: warning: This header is deprecated. Use <boost/iterator/function_output_iterator.hpp> instead. [-W#pragma-messages]
BOOST_HEADER_DEPRECATED("<boost/iterator/function_output_iterator.hpp>")
^
/usr/local/include/boost/config/header_deprecated.hpp:23:37: note: expanded from macro 'BOOST_HEADER_DEPRECATED'
# define BOOST_HEADER_DEPRECATED(a) BOOST_PRAGMA_MESSAGE("This header is deprecated. Use " a " instead.")
                                    ^
/usr/local/include/boost/config/pragma_message.hpp:24:34: note: expanded from macro 'BOOST_PRAGMA_MESSAGE'
# define BOOST_PRAGMA_MESSAGE(x) _Pragma(BOOST_STRINGIZE(message(x)))
                                 ^
<scratch space>:21:2: note: expanded from here
 message("This header is deprecated. Use " "<boost/iterator/function_output_iterator.hpp>" " instead.")
 ^
2 warnings and 2 errors generated.
make[2]: *** [lib/base/CMakeFiles/base.dir/application-environment.cpp.o] Error 1
In file included from /Users/virt/Desktop/icinga2/lib/base/array.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/array.hpp:6:
In file included from /Users/virt/Desktop/icinga2/lib/base/i2-base.hpp:38:
In file included from /usr/local/include/boost/config.hpp:44:
/usr/local/include/boost/config/detail/select_stdlib_config.hpp:24:14: warning: non-portable path to file '<VERSION>'; specified path differs in case from file name on disk [-Wnonportable-include-path]
#    include <version>
             ^~~~~~~~~
             <VERSION>
In file included from /Users/virt/Desktop/icinga2/lib/base/array.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/array.hpp:6:
In file included from /Users/virt/Desktop/icinga2/lib/base/i2-base.hpp:38:
In file included from /usr/local/include/boost/config.hpp:44:
In file included from /usr/local/include/boost/config/detail/select_stdlib_config.hpp:24:
/Users/virt/Desktop/icinga2/version:1:1: error: unknown type name 'Version'
Version: 2.12.0
^
/Users/virt/Desktop/icinga2/version:1:8: error: expected unqualified-id
Version: 2.12.0
       ^
In file included from /Users/virt/Desktop/icinga2/lib/base/array.cpp:10:
In file included from /Users/virt/Desktop/icinga2/lib/base/exception.hpp:12:
In file included from /Users/virt/Desktop/icinga2/lib/base/configobject.hpp:7:
In file included from /Users/virt/Desktop/icinga2/debug/lib/base/configobject-ti.hpp:9:
In file included from /usr/local/include/boost/signals2.hpp:19:
In file included from /usr/local/include/boost/signals2/signal.hpp:22:
In file included from /usr/local/include/boost/signals2/connection.hpp:22:
In file included from /usr/local/include/boost/signals2/detail/null_output_iterator.hpp:14:
/usr/local/include/boost/function_output_iterator.hpp:14:1: warning: This header is deprecated. Use <boost/iterator/function_output_iterator.hpp> instead. [-W#pragma-messages]
BOOST_HEADER_DEPRECATED("<boost/iterator/function_output_iterator.hpp>")
^
/usr/local/include/boost/config/header_deprecated.hpp:23:37: note: expanded from macro 'BOOST_HEADER_DEPRECATED'
# define BOOST_HEADER_DEPRECATED(a) BOOST_PRAGMA_MESSAGE("This header is deprecated. Use " a " instead.")
                                    ^
/usr/local/include/boost/config/pragma_message.hpp:24:34: note: expanded from macro 'BOOST_PRAGMA_MESSAGE'
# define BOOST_PRAGMA_MESSAGE(x) _Pragma(BOOST_STRINGIZE(message(x)))
                                 ^
<scratch space>:114:2: note: expanded from here
 message("This header is deprecated. Use " "<boost/iterator/function_output_iterator.hpp>" " instead.")
 ^
2 warnings and 2 errors generated.
make[2]: *** [lib/base/CMakeFiles/base.dir/array.cpp.o] Error 1
In file included from /Users/virt/Desktop/icinga2/lib/base/application.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/application.hpp:6:
In file included from /Users/virt/Desktop/icinga2/lib/base/i2-base.hpp:38:
In file included from /usr/local/include/boost/config.hpp:44:
/usr/local/include/boost/config/detail/select_stdlib_config.hpp:24:14: warning: non-portable path to file '<VERSION>'; specified path differs in case from file name on disk [-Wnonportable-include-path]
#    include <version>
             ^~~~~~~~~
             <VERSION>
In file included from /Users/virt/Desktop/icinga2/lib/base/application.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/application.hpp:6:
In file included from /Users/virt/Desktop/icinga2/lib/base/i2-base.hpp:38:
In file included from /usr/local/include/boost/config.hpp:44:
In file included from /usr/local/include/boost/config/detail/select_stdlib_config.hpp:24:
/Users/virt/Desktop/icinga2/version:1:1: error: unknown type name 'Version'
Version: 2.12.0
^
/Users/virt/Desktop/icinga2/version:1:8: error: expected unqualified-id
Version: 2.12.0
       ^
In file included from /Users/virt/Desktop/icinga2/lib/base/application.cpp:3:
In file included from /Users/virt/Desktop/icinga2/lib/base/application.hpp:7:
In file included from /Users/virt/Desktop/icinga2/debug/lib/base/application-ti.hpp:9:
In file included from /usr/local/include/boost/signals2.hpp:19:
In file included from /usr/local/include/boost/signals2/signal.hpp:22:
In file included from /usr/local/include/boost/signals2/connection.hpp:22:
In file included from /usr/local/include/boost/signals2/detail/null_output_iterator.hpp:14:
/usr/local/include/boost/function_output_iterator.hpp:14:1: warning: This header is deprecated. Use <boost/iterator/function_output_iterator.hpp> instead. [-W#pragma-messages]
BOOST_HEADER_DEPRECATED("<boost/iterator/function_output_iterator.hpp>")
^
/usr/local/include/boost/config/header_deprecated.hpp:23:37: note: expanded from macro 'BOOST_HEADER_DEPRECATED'
# define BOOST_HEADER_DEPRECATED(a) BOOST_PRAGMA_MESSAGE("This header is deprecated. Use " a " instead.")
                                    ^
/usr/local/include/boost/config/pragma_message.hpp:24:34: note: expanded from macro 'BOOST_PRAGMA_MESSAGE'
# define BOOST_PRAGMA_MESSAGE(x) _Pragma(BOOST_STRINGIZE(message(x)))
                                 ^
<scratch space>:21:2: note: expanded from here
 message("This header is deprecated. Use " "<boost/iterator/function_output_iterator.hpp>" " instead.")
 ^
2 warnings and 2 errors generated.
make[2]: *** [lib/base/CMakeFiles/base.dir/application.cpp.o] Error 1
make[1]: *** [lib/base/CMakeFiles/base.dir/all] Error 2
make: *** [all] Error 2
```

## To Reproduce

1. git clone https://github.com/Icinga/icinga2.git
2. cd icinga2
3. mkdir debug
4. cd debug
5. export CC='ccache clang'; export CXX='ccache clang++'; export CCACHE_CPP2=yes
6. cmake \
-DCMAKE_BUILD_TYPE=Debug \
-DICINGA2_UNITY_BUILD=OFF \
-DICINGA2_WITH_MYSQL=OFF \
-DICINGA2_WITH_PGSQL=OFF \
-DCMAKE_INSTALL_PREFIX=/usr/local/icinga2 \
-DICINGA2_PLUGINDIR=/usr/local/sbin \
-DOPENSSL_INCLUDE_DIR=/usr/local/opt/openssl@1.1/include \
-DBUILD_TESTING=OFF \
-DOPENSSL_SSL_LIBRARY=/usr/local/opt/openssl@1.1/lib/libssl.dylib \
-DOPENSSL_CRYPTO_LIBRARY=/usr/local/opt/openssl@1.1/lib/libcrypto.dylib \
..
7. sudo make -j4 install -C debug

## Expected behavior

It should compile? At least it did for quite some time...

**Comment by Al2Klimov** [comment]
@Yonas-net Have you ever seen this one?

**Comment by Yonas-net** [comment]
Oh yes, after running brew upgrade once I couldn't build Icinga2, I got exactly the same error as above and it still doesn't work even with freshly installed boost 1.66, 172, 173 libraries.

**Comment by Al2Klimov** [comment]
And we didn’t find a solution for it?

**Comment by Yonas-net** [comment]
Yes, for me it was, after you sent me the ready installed and built 166 boost library.

**Comment by Al2Klimov** [comment]
@henningkessler Please try to build Boost from source, install it into a custom location and use it for Icinga.

**Comment by julianbrost** [comment]
What's the current state of file-systems on macOS regarding case-sensitivity?

Does `brew` (I assume Boost installed from there) patch Boost in some way? My Boost 1.72 does not try to include `<version>` (technically not fully correct, but only in `boost/multiprecision`).

**Comment by Al2Klimov** [comment]
1. Case-insensitive. (I had to reformat my HDD to check out Linux via Git.)
2. I guess not as this issue is there.

**Comment by julianbrost** [comment]
So there is https://github.com/boostorg/config/commit/26f20bd8e5e8ca7b41f24ff87e4c8c6ce24d29b4 which is a problem in combination with our `VERSION` file on case-insensitive file systems. But that's only in Boost 1.73 and later. I know that building with Boost 1.72 fails unless you apply the patch listed under known issues at https://www.boost.org/users/history/version_1_72_0.html, but I don't see why 1.66 should fail.

**Comment by henningkessler** [comment]
I can confirm that compiling Icinga works with a manual compiled  Boost version 1.72 (including the patch). My Problem therefor was Boost version 1.74 which was the default version from Homebrew.

**Comment by Al2Klimov** [comment]
To be honest MasOS isn't an officially supported platform, neither a platform Icinga is expected to run on _in production_. If someone (incl. devs) builds Icinga on a such platform, IMAO all obstacles are their problem. I'll let this open, but I consider this low prio.

**Comment by Al2Klimov** [comment]
Fixed by #8596.
